# Laboratorio 2 - Optimización

### Isabella Salgado 201730418
### Juan Pablo Naranjo 201730006

Se importan las librerías relevantes para el laboratorio.

In [1]:
import numpy as np 
import matplotlib.pyplot as mat
import math
import time
from mpl_toolkits import mplot3d
from mpl_toolkits.mplot3d import Axes3D
import sympy as sym
from scipy.optimize import linprog
from scipy.optimize import minimize

## Punto 1

## a) 
Las variables de decisión son el tamaño de los productos: pequeño, mediano y grande. Estas se representan a través de las variables $x_1$ (grande), $x_2$ (mediano) y $x_3$ (pequeño).

Se define la notación $x_{ij}$ para expresar los productos que pertenecen a la planta $i$, del tamaño $j$.

Adicionalmente, la cantidad de productos ($x_j$) están definidas por las siguientes ecuaciones:

1) $x_1 = x_{11} + x_{21} + x_{31}$ (Grande)

2) $x_2 = x_{12} + x_{22} + x_{32}$ (Mediano)

3) $x_3 = x_{13} + x_{23} + x_{33}$ (Pequeño)


## b) 
La función objetivo está dada por $f(x)$. 

$f(x) = 140x_1 + 120x_2 + 100x_3$ \
$f(x) = 140(x_{11} + x_{21} + x_{31}) + 120(x_{12} + x_{22} + x_{32}) + 100(x_{13} + x_{23} + x_{33})$ \
$f(x) = 140x_{11} + 140x_{21} + 140x_{31} +  120x_{12} + 120x_{22} + 120x_{32} + 100x_{13} + 100x_{23} + 100x_{33}$

### A continuación se presentan las restricciones de espacio en las plantas:

Planta 1:

$20x_{11} + 15x_{12} + 12x_{13} \leq 13000$

Planta 2:

$20x_{21} + 15x_{22} + 12x_{23} \leq 12000$

Planta 3:

$20x_{31} + 15x_{32} + 12x_{33} \leq 5000$

### A continuación se presentan las restricciones por capacidad de producción de las plantas:

Planta 1:

$x_{11} + x_{12} + x_{13} \leq 750$

Planta 2:

$x_{21} + x_{22} + x_{23} \leq 900$

Planta 3:

$x_{31} + x_{32} + x_{33} \leq 450$

### A continuación se presntan las restricciones por los pronósticos de venta:

Grande:

$x_1 = x_{11} + x_{21} + x_{31} \leq 900$

Mediano:

$x_2 = x_{12} + x_{22} + x_{32} \leq 1200$

Pequeño:

$x_3 = x_{13} + x_{23} + x_{33} \leq 750$

## c) Resolución del problema de optimización 

Se colocan los valores del vector negativos porque se está minimizando el negativo de la función objetivo.

In [3]:
c = np.array([-140, -120, -100, -140, -120, -100, -140, -120, -100])
A = np.array([[20,15,12,0,0,0,0,0,0], [0,0,0,20,15,12,0,0,0], [0,0,0,0,0,0,20,15,12], [1,1,1,0,0,0,0,0,0], [0,0,0,1,1,1,0,0,0], [0,0,0,0,0,0,1,1,1], [1,0,0,1,0,0,1,0,0], [0,1,0,0,1,0,0,1,0], [0,0,1,0,0,1,0,0,1]])
b = np.array([13000, 12000, 5000, 750, 900, 450, 900, 1200, 750])
x11_bounds = (0, None)
x12_bounds = (0, None)
x13_bounds = (0, None)
x21_bounds = (0, None)
x22_bounds = (0, None)
x23_bounds = (0, None)
x31_bounds = (0, None)
x32_bounds = (0, None)
x33_bounds = (0, None)

res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x11_bounds, x12_bounds, x13_bounds, x21_bounds, x22_bounds, x23_bounds, x31_bounds, x32_bounds, x33_bounds]), method='revised simplex', options={"disp": True})
print(res)

Phase Iteration Minimum Slack       Constraint Residual Objective          
1     0         450.0               0.0                 0.0                 
Phase Iteration Minimum Slack       Constraint Residual Objective          
2     0         450.0               0.0                 0.0                 
2     1         0.0                 0.0                 -91000.0            
2     2         0.0                 0.0                 -126000.0           
2     3         0.0                 0.0                 -174000.0           
2     4         0.0                 0.0                 -182000.0           
2     5         0.0                 0.0                 -222000.0           
2     6         0.0                 0.0                 -228000.0           
2     7         0.0                 0.0                 -234666.6666667     
2     8         0.0                 0.0                 -236000.0           
Optimization terminated successfully.
         Current function value: -236000

### Valor de la función objetivo:

In [4]:
print(-1*res.fun)

236000.0


### Valor de cada una de las variables:

In [5]:
print(res.x)

[350.         400.           0.           0.         533.33333333
 333.33333333   0.           0.         416.66666667]


### Se deben producir:

**350** productos de tamaño **grande** en la planta 1 ($x_{11} = 350$)

**400** productos de tamaño **mediano** en la planta 1 ($x_{12} = 400$)

**0** productos de tamaño **pequeño** en la planta 1 ($x_{13} = 0$)

### Dando un total de producción de en la planta 1 de 

$x_1 = x_{11} + x_{12} + x_{13} = 750$ productos

### Se deben producir:

**0** productos de tamaño **grande** en la planta 2 ($x_{21} = 0$)

**533.33** productos de tamaño **mediano** en la planta 2 ($x_{22} = 533.33$)

**333.33** productos de tamaño **pequeño** en la planta 2 ($x_{23} = 333.33$)

### Dando un total de producción en la planta 2 de

$x_2 = x_{21} + x_{22} + x_{23} = 866.66$ productos

*Nota: Ya que en la realidad no se puede producir un número irracional de productos, se redondearía este valor al entero más cercano, que en este caso sería 867*

### Se deben producir:

**0** productos de tamaño **grande** en la planta 3 ($x_{31} = 0$)

**0** productos de tamaño **mediano** en la planta 3 ($x_{32} = 0$)

**416.66** productos de tamaño **pequeño** en la planta 3 ($x_{33} = 416.66$)

### Dando un total de producción en la planta 3 de

$x_3 = x_{31} + x_{32} + x_{33} = 416.66$ productos

*Nota: Ya que en la realidad no se puede producir un número irracional de productos, se redondearía este valor al entero más cercano, que en este caso sería 417*

**Estos resultados garantizan que la utilidad se maximice en un valor de $f(x) = \$236,000$**

## Punto 2

### a) Definición de las variables de decisión
Posición $x$ (horizontal) \
Posición $y$ (vertical) \
De la bodega en el plano cartesiano.

### b) Función objetivo
Distancia entre el local A y la bodega: $d_{a} = \sqrt{x^{2}+y^{2}}$ 

Distancia entre el local B y la bodega: $d_{b} = \sqrt{(x-40)^{2}+(y-50)^{2}}$ 

Distancia entre el local C y la bodega: $d_{c} = \sqrt{(x-90)^{2}+(y-30)^{2}}$

Función a optimizar: $f(x,y) = d_a + d_b + d_c$

In [6]:
def f(x):
    return sym.sqrt(x[0]**2+x[1]**2) + sym.sqrt((x[0]-40)**2 + (x[1]-50)**2) + sym.sqrt((x[0]-90)**2+(x[1]-30)**2)

### c) ¿Se puede reformular este problema de manera lineal sin cambiar las variables de decisión?

Este problema no se puede reformular de manera lineal sin cambiar las variables de decisión porque para lograr esto tocaría hacer un cambio de variable que implicaría una transformación matemática compleja para eliminar las raíces cuadradas que de todas formas no garantiza la linealidad del problema. Además, hacer esto implica un cambio en las variables de decisión.

### d) Optimización por el método de Broyden-Fletcher-Goldfarb-Shanno

In [7]:
# Se define un vector de valores iniciales de las variables de decisión.
x0 = [0,0]

res = minimize(f, x0, method='bfgs', options={'disp': True})
print(res)
print(np.round(res.x,3))

Optimization terminated successfully.
         Current function value: 117.114336
         Iterations: 11
         Function evaluations: 48
         Gradient evaluations: 16
      fun: 117.11433586462294
 hess_inv: array([[ 9.29048917, -7.63334816],
       [-7.63334816, 33.88074963]])
      jac: array([ 1.90734863e-06, -2.86102295e-06])
  message: 'Optimization terminated successfully.'
     nfev: 48
      nit: 11
     njev: 16
   status: 0
  success: True
        x: array([42.04481069, 42.54560226])
[42.045 42.546]


### e)
De acuerdo al resultado obtenido en el apartado $d)$, se puede afirmar que la ubicación óptima para la bodega (más cercana posible a los locales A, B y C) corresponde a las coordenadas ($x$, $y$) = (42.045, 42.546).

## Punto 3

### a) Expresión matemática

Distancia entre el local A y su respectiva bodega: $d_{a} = \sqrt{x^{2}+y^{2}}$ 

Distancia entre el local B y su respectiva bodega: $d_{b} = \sqrt{(x-40)^{2}+(y-50)^{2}}$ 

Distancia entre el local C y su respectiva bodega: $d_{c} = \sqrt{(x-90)^{2}+(y-30)^{2}}$

Se definen las variables $x_1$, $y_1$ para representar las coordenadas de la bodega correspondiente al local A.

Se definen las variables $x_2$, $y_2$ para representar las coordenadas de la bodega correspondiente al local C.

Función a optimizar:

$f(x_1,x_2,y_1,y_2) = \sqrt{x_{1}^{2}+y_{1}^{2}} + \frac{1}{2} \left(\sqrt{(x_{1}-40)^{2}+(y_{1}-50)^{2}} + \sqrt{(x_{2}-40)^{2}+(y_{2}-50)^{2}}\right) + \sqrt{(x_{2}-90)^{2} + (y_{2}-30)^{2}} $

In [16]:
# x es un vector de 4 componentes. Cada componente representa una de las variables de la función objetivo.
def f1(x):
    return np.sqrt(x[0]**2+x[1]**2) + (1/2)*(np.sqrt((x[0]-40)**2 + (x[1]-50)**2) + np.sqrt((x[2]-40)**2 + (x[3]-50)**2)) + np.sqrt((x[2]-90)**2+(x[3]-30)**2)
    

### b) Minimización de la función objetivo

Información sobre cómo usar la función *minimize* se encontró [acá](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html).

In [20]:
x1 = [0,0,0,0]

res2 = minimize(f1, x1, method='BFGS', options={'disp': True})
print(res2)

         Current function value: 58.941445
         Iterations: 65
         Function evaluations: 841
         Gradient evaluations: 166
      fun: 58.94144523431122
 hess_inv: array([[ 1.61908938e-09, -2.64955636e-09, -1.74805267e-09,
         3.42094131e-10],
       [-2.64955636e-09,  6.53495986e-08,  1.30949709e-08,
         2.01647431e-09],
       [-1.74805267e-09,  1.30949709e-08,  1.11834206e-08,
         6.97415740e-11],
       [ 3.42094131e-10,  2.01647431e-09,  6.97415740e-11,
         8.48260527e-09]])
      jac: array([-1.30318689,  0.56361675,  0.43057299, -0.33080721])
  message: 'Desired error not necessarily achieved due to precision loss.'
     nfev: 841
      nit: 65
     njev: 166
   status: 2
  success: False
        x: array([-1.19611787e-09,  6.13580996e-10,  9.00000000e+01,  3.00000000e+01])


El resultado obtenido para las primeras dos coordenadas ($x_1,y_1$) son números en el orden de $10^{-9}$ y $10^{-10}$, lo que quiere decir que esos números son básicamente 0.

Como se puede notar, la respuesta obtenida para la ubicación de las bodegas son respectivamente en las ubicaciones de los locales A y C, es decir, en las coordenadas (0, 0) para el local A, y en las coordenadas (90, 30) para el local C.

In [18]:
res21 = minimize(f1, x1, method='Nelder-Mead', options={'disp': True})
print(res21)

Optimization terminated successfully.
         Current function value: 82.696593
         Iterations: 216
         Function evaluations: 360
 final_simplex: (array([[-14.23296298,  -7.69426269,  89.99996647,  29.99999978],
       [-14.2329666 ,  -7.69423354,  89.99997058,  29.99995387],
       [-14.23296138,  -7.69421418,  89.99992764,  29.99991446],
       [-14.2329755 ,  -7.69423101,  90.00002268,  29.99996013],
       [-14.23294723,  -7.69425905,  89.99987015,  29.99997513]]), array([82.69659292, 82.69660447, 82.69662655, 82.69662741, 82.69662915]))
           fun: 82.6965929183624
       message: 'Optimization terminated successfully.'
          nfev: 360
           nit: 216
        status: 0
       success: True
             x: array([-14.23296298,  -7.69426269,  89.99996647,  29.99999978])


En este caso, se utilizó el método de Nelder-Mead en lugar del método BFGS para llevar a cabo la minimización. Como se puede notar, el resultado obtenido para las coordenadas de la primera bodega ($x_1, y_1$) están muy alejadas de las obtenidas por el método BFGS. Mientras tanto, las coordenadas de la segunda bodega ($x_2, y_2$) sí son similares a las obtenidas por el método BFGS, estando alrededor del punto (90, 30) nuevamente.

Esto se puede deber al hecho de que el algoritmo de [Nelder-Mead](http://www.scholarpedia.org/article/Nelder-Mead_algorithm) solo usa valores de la función objetivo en algunos puntos de $\mathbb{R}^n$ y no calcula el gradiente de la función objetivo en estos puntos para dirigirse en la dirección de la solución óptima, como si lo hace el método [BFGS](https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93Goldfarb%E2%80%93Shanno_algorithm). Esto puede estar causando la discrpeancia entre los valores obtenidos para las coordenadas ($x_1,y_1$) entre los dos métodos analizados.

### c) Según los resultados anteriores, ¿cuál de las dos opciones (entre 1 y 2 bodegas) le recomendaría usted al gerente de la empresa?

De acuerdo a los resultados obtenidos en el punto 2 y en el apartado *b)* de este punto, se puede afirmar que es más recomendable tener dos bodegas distintas (una para el local A y uno para el C, distribuyéndose la producción del local B) porque de este modo, se obtiene que la ubicación óptima para ambas bodegas es justo al lado de su rspectivo local, pues es la distancia más corta posible que hay entre cada una y su respectivo local.

Con una sola bodega para los tres locales se obtiene una ubicación óptima de la bodega justo en un punto relativamente medio para los tres locales, aunque es claro que la distancia al local A es más alejada que la distancia a los otros dos locales respectivamente. 